### PR TITLE
Cypher: array access account for sorted edges

### LINKED ISSUE #230: Sort arrow property arrays when sorting edges
The Cypher graph data structure sorts edges by destination/label for efficiency. The arrow property arrays for the edges are not sorted along with them at the moment, leading to a mismatch between an edge and its property. It causes things like this:

```
||  (:CONTAINER_OF  {} )  ||
||  (:CONTAINER_OF  {joinDate: 2017-08-28T06:24:31.056Z,  } ) ||
||  (:CONTAINER_OF  {} )  ||
||  (:CONTAINER_OF  {} )  ||
||  (:CONTAINER_OF  {joinDate: 2013-06-07T01:11:52.767Z,  } ) ||
||  (:CONTAINER_OF  {joinDate: 2014-12-12T20:13:48.712Z,  } ) ||
||  (:CONTAINER_OF  {} )  ||
```

In LDBC data, CONTAINER_OF edge types should not have any attribute attached with them, but in this case, they do.

This can be resolved by either sorting the arrow property arrays along with the edge destinations or by mapping the resorted edge index to the original index in order to index into the array correctly. I am currently working on the former, but if there are objections, this is the place to discuss them.

### PR DESCRIPTION
**Pull request by l-hoang**
Before this commit, arrow array accesses for edges were incorrect as the
underlying CSR graph is sorted for efficiency while the arrow arrays are
not sorted along with them. This causes references to a particular edge
index to be incorrect. This commit adds a mapping from original edge
index to new edge index that's used when accessing the arrow arrays; it
is constructed by sorting a 1-to-1 mapping in the same manner as the CSR
is sorted.

The original plan was to sort all of the arrow arrays to avoid a layer
of indirection for each edge property access, but complications there
included creating a new arrow chunked array for the sorted array which
included having to deal with null elements during the sort. This is
not to say this isn't doable; in fact, the original to new edge mapping
created in this commit can be used to facilitate the creation of new
arrow arrays for this purpose. That can be done in another PR.

Resolves #230

### CODE REVIEW

File: libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h [hunk 1]
```diff
@@ -511,6 +511,48 @@ class LC_CSR_Labeled_Graph
...
 
+  /**
+   * Given some vector, sort the indices of that vector as if they were
+   * the edge destinations that would get sorted if one sorted by the
+   * edge data then destinations. Used mainly to rearrange other vectors
+   * that also need to be sorted besides destinations/data  (since the
+   * current infrastructure only supports sorting those 2 arrays at
+   * the moment).
+   */
+  void sortVectorByDataThenDst(std::vector<uint64_t>& vector_to_sort) {
```
Comment (roshandathathri) [inline libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:522]: I think this should be a private function and the vector should be a member variable. We should have a function which takes an iterator and returns an iterator. This function should map the edge from old index to new index.

Comment (roshandathathri) [reply libgalois/include/galois/graphs/LC_CSR_Labeled_Graph.h:522]: Also, we would two such private functions and two vector member variables - one for outgoing edges and one for incoming edges.

File: libquery/include/querying/IndexableChunkedArray.h [hunk 2]
```diff
@@ -66,6 +76,11 @@ class IndexableChunkedArray {
   getNumericElement(int64_t global_index) const {
     // user is responsible for correct numeric type outside of debug build
     assert(this->getType()->Equals(NumType()));
+    assert((int64_t)this->index_to_sorted_edge.size() == this->size());
+
+    // get the remapped sorted index for this edge
+    global_index = this->index_to_sorted_edge[global_index];
```
Comment (roshandathathri) [inline libquery/include/querying/IndexableChunkedArray.h:82]: This should be done outside the arrow interface.

File: libquery/src/IndexableChunkedArray.cpp [hunk 3]
```diff
@@ -68,6 +70,11 @@ std::optional<std::string>
 IndexableChunkedArray::getStringElement(int64_t global_index) const {
   // user is responsible for calling appropriate function outside debug
   assert(this->getType()->Equals(arrow::utf8()));
+  assert((int64_t)this->index_to_sorted_edge.size() == this->size());
+
+  // get the remapped sorted index for this edge
+  global_index = this->index_to_sorted_edge[global_index];
```
Comment (roshandathathri) [inline libquery/src/IndexableChunkedArray.cpp:76]: Ditto.

**Comment by l-hoang** [comment]
re: outgoing/incoming; I haven't thought about how to do incoming, but what you said makes sense. I haven't thought about this much, but I think you can get away with just one remapping array since the incoming -> outgoing might already exist from LC_CSR_CSC, no? (unless we aren't using that feature)

Also agreed about moving the reindexing out of the arrow array, especially since as you mentioned I'm not considering incoming edges yet and that would screw up how that works if it's kept as is.

In any case, isn't the idea to abandon CSR_Labeled graph soon anyways for PropertyGraph? (as in I hope the next week or so :P)

**Pull request merged**